### DISCUSSION TITLE
Removing method sending from intermediate languages

### DISCUSSION THREAD

**Discussion by Ekdohibs** [ocaml/ocaml#12124] [General]
In lambda and clambda (and flambda as well), we have a "send" construction used for method sending, which takes an object, a method and some arguments, extracts the method from the object, and then calls it with the object and the arguments.

We (the flambda2 team) propose to remove this special construct, simply replacing it by a normal application, and a single primitive which computes the method for an object. This would greatly simplify the treatment of method calls, by removing this special case from the intermediate languages, as well as allowing to remove all the `caml_send` functions. Besides, it allows new optimisation opportunities: the computation of a method for a given object can be shared by CSE.

We implemented this in flambda2, as can be seen in this pull request: https://github.com/ocaml-flambda/flambda-backend/pull/1213
Some preliminary micro-benchmarks show that there is a performance impact on method calls, but it seems to be limited: on the order of 10-20% on a micro-benchmark containing a single method call in a loop, where the method does extremely little work, as shown below:
```ocaml
let o = object
  val mutable x = 0
  method f i = x <- x + i; x
end

let n =
  (* ignore (o#f); *)
  for i = 0 to 10000000000 do
    ignore (o#f i)
  done;
  o#f 0
```
Besides, in this case, if we uncomment the `ignore (o#f)`, common subexpression elimination is able to share all the computations of methods, showing a performance improvement on the order of 10-20% (note that this is an improvement even though the method exists in the cache and is found immediately).

In the flambda2 team, we would be very interested to simplify this part of the compiler. As this is something orthogonal to what we usually work on, it would be quite simple to upstream, so we can do that if there is interest for it. However, as we almost do not use objects at all, we do not know if the performance impact and benefits would matter. Thus, would there be interest in simplifying this?

**Comment by nojb** [comment]
> on the order of 10-20% on a micro-benchmark

Do you understand where this slowdown comes from?

**Reply by Ekdohibs** [reply]
Unfortunately, only partially. Some of the slowdown seems to be caused by the additional calls that happens: while upstream, the computation of the method is inlined inside the different `caml_send` functions, here we do not want to inline it at every call site and we end up doing two calls, one to get the method, and one to apply it (it has one argument, so with the object, we need to call `caml_apply2`). As the loop is very small (only 37 instructions executed for each iteration upstream, 41 when removing send), this difference probably explains part of the slowdown.
However, there seem to be other causes which we have not been able to identify: for instance, depending on whether `caml_get_cached_public_method` is implemented in C or in cmm code, we get an additional performance difference (C being around 10% slower than cmm), and we have not been able to explain this difference.
